src/dbnode/persist/fs/commitlog/commit_log.go

```diff
@@ -506,69 +541,169 @@ func (l *commitLog) write() {
 		l.metrics.success.Inc(numWritesSuccess)
 	}
 
-	writer := l.writerState.writer
-	l.writerState.writer = nil
+	// Ensure that there is no active background goroutine in the middle of reseting
+	// the secondary writer / modifying its state.
+	l.waitForSecondaryWriterAsyncResetComplete()
+	// Can be nil in the case where the background goroutine spawned in openWriters
+	// encountered an error trying to re-open it.
+	if l.writerState.secondaryWriter.writer != nil {
+		// Don't care about errors closing the secondary writer because it doesn't
+		// have any data.
+		l.writerState.secondaryWriter.writer.Close()
+		l.writerState.secondaryWriter.writer = nil
+	}
+
+	writer := l.writerState.primaryWriter.writer
+	l.writerState.primaryWriter.writer = nil
```

Comment:
Instead of the above, if you have the array of writers it gets a little eaiser:

```go
var multiErr xerrors.MultiError
for _, writer := l.writerState.writers {
  multiErr = multiErr.Add(writer.Close())
}

// Zero our the writers
l.writerState = writerState{}

// Report close error
l.closeErr <- multiErr.FinalError()
```